### PR TITLE
[Feature] qwen3.5 support pcp+mtp

### PR DESCRIPTION
**Pull request by SunnyLee151064**
### What this PR does / why we need it?
This PR introduces support for hybrid attention PCP (Parallel Context Processing) by adding linear-splitting of prefill inputs (`_split_pcp_input_hybrid`) and updating the model runner to handle hybrid attention logits indices. It also corrects several index offsets in `pcp_utils.py` (switching between decode requests and decode tokens) and updates Triton operations for Mamba and FLA chunk-gated delta rules.

However, several critical issues were identified during the review:
- In `causal_conv1d.py`, `num_prefills` is initialized to `0` but never updated from `attn_metadata`, which breaks the PCP + MTP logic.
- In `model_runner_v1.py`, calculating `num_scheduled_tokens` via slice differences misses the first request's token count, and an out-of-bounds/incorrect index access occurs when `num_decode_reqs` is `0`.
- In `chunk.py`, calling `.item()` on an NPU tensor triggers an expensive device-to-host synchronization in the hot path.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
No testing details were provided in the patch.

- vLLM version: v0.22.1
- vLLM main: https://github.com/vllm-project/vllm/commit/967c5c3bc38891f4465d3f4e99917ed837bb3833

### CODE REVIEW

File: vllm_ascend/spec_decode/llm_base_proposer.py [hunk 1]
```diff
@@ -1355,7 +1355,7 @@ def set_inputs_first_pass(
             if self.uses_xdrope_dim > 0 and self.draft_uses_xdrope_dim == 0:
                 target_positions = target_positions[0]
 
-            self._set_positions(num_tokens, target_positions)
+            self._set_positions(target_positions.shape[-1], target_positions)
```
Comment (lilinsiman) [inline vllm_ascend/spec_decode/llm_base_proposer.py]: It is no need to change here

File: vllm_ascend/worker/model_runner_v1.py [hunk 2]
```diff
@@ -1564,11 +1567,29 @@ def _calc_spec_decode_metadata(
...
         if self.pcp_size > 1:
-            cu_num_scheduled_tokens = cu_num_scheduled_tokens * self.pcp_size - num_pcp_pads
+            if self.pcp_manager.pcp_use_hybrid_attn: 
+                if self.pcp_manager.num_prefill_reqs > 0:
+                    num_scheduled_tokens = cu_num_scheduled_tokens[1:] - cu_num_scheduled_tokens[:-1]
+                    prefill_num_scheduled_tokens = num_scheduled_tokens[-self.pcp_manager.num_prefill_reqs: ]
+                    prefill_num_scheduled_tokens = [math.ceil(num / (self.pcp_size * 2)) * 2 * self.pcp_size 
+                                                    for num in prefill_num_scheduled_tokens]
+                    prefill_num_scheduled_tokens = list(accumulate(prefill_num_scheduled_tokens))
+                    prefill_num_scheduled_tokens = [cu_num_scheduled_tokens[self.pcp_manager.num_decode_reqs - 1] + num
```
Comment (Tflowers-0129) [inline vllm_ascend/worker/model_runner_v1.py]: 越界行为可能发生

File: vllm_ascend/worker/model_runner_v1.py [hunk 3]
```diff
@@ -1564,11 +1567,29 @@ def _calc_spec_decode_metadata(
...
+                                                    for num in prefill_num_scheduled_tokens]
+                    cu_num_scheduled_tokens[-self.pcp_manager.num_prefill_reqs:] = prefill_num_scheduled_tokens
+                    cu_num_scheduled_tokens = cu_num_scheduled_tokens.copy()
+                    cu_num_scheduled_tokens[self.pcp_manager.num_decode_reqs:] = (
+                        cu_num_scheduled_tokens[self.pcp_manager.num_decode_reqs:] * self.pcp_size 
+                        - num_pcp_pads[self.pcp_manager.num_decode_reqs:]) # type: ignore
+
+            else:
+                cu_num_scheduled_tokens = cu_num_scheduled_tokens * self.pcp_size - num_pcp_pads
             logits_indices_pcp = np.repeat(cu_num_scheduled_tokens - num_sampled_tokens, num_sampled_tokens)
```
Comment (Tflowers-0129) [inline vllm_ascend/worker/model_runner_v1.py]: 增加UT

File: vllm_ascend/spec_decode/llm_base_proposer.py [hunk 4]
```diff
@@ -1999,6 +2002,47 @@ def _pcp_pad_and_split(num_tokens):
         cu_num_tokens = torch.tensor(np.insert(np.cumsum(np.array(num_pcp_scheduled_tokens)), 0, 0))
         return num_tokens, input_ids, target_hidden_states, max_query_len, seq_lens, cu_num_tokens
 
+    def _split_pcp_input_hybrid(self, req_scheduled_tokens, input_ids, target_hidden_states):
```
Comment (Tflowers-0129) [inline vllm_ascend/spec_decode/llm_base_proposer.py:2004]: 增加UT

File: vllm_ascend/spec_decode/llm_base_proposer.py [hunk 5]
```diff
@@ -1999,6 +2002,47 @@ def _pcp_pad_and_split(num_tokens):
...
+
+        Example: ori_num_tokens=7, pcp_size=2 → padded=8, pcp_tokens=4
+          rank 0: [0,1,2,3]  (4 valid)
+          rank 1: [4,5,6,7]  (3 valid + 1 pad at position 7)
+        """
+        num_pcp_scheduled_tokens = []
+        global_offset = 0
+        pcp_split_input_ids_list = []
+        pcp_split_hidden_states_list = []
+        for ori_num_tokens in req_scheduled_tokens.values():
```
Comment (Tflowers-0129) [inline vllm_ascend/spec_decode/llm_base_proposer.py:2020]: 如果 req_scheduled_tokens 为空，会直接报错。

File: vllm_ascend/ops/triton/fla/chunk.py [hunk 6]
```diff
@@ -151,7 +151,12 @@ def chunk_gated_delta_rule_fwd(
 
         if get_pcp_group().rank_in_group > 0:
             rerun_initial_state = initial_state.clone()
-            prefill_slice = slice(num_decodes, final_state.shape[0])
+            if cu_seqlens is not None:
+                _ns_lens = cu_seqlens[1:] - cu_seqlens[:-1]
+                prefill_seq_offset = int(((_ns_lens > 0) & (_ns_lens <= 1)).sum().item())
```
Comment (weiguihua2) [inline vllm_ascend/ops/triton/fla/chunk.py:156]: item will be synchronous operations.

File: vllm_ascend/spec_decode/llm_base_proposer.py [hunk 7]
```diff
@@ -2001,6 +2004,47 @@ def _pcp_pad_and_split(num_tokens):
         cu_num_tokens = torch.tensor(np.insert(np.cumsum(np.array(num_pcp_scheduled_tokens)), 0, 0))
         return num_tokens, input_ids, target_hidden_states, max_query_len, seq_lens, cu_num_tokens
 
+    def _split_pcp_input_hybrid(self, req_scheduled_tokens, input_ids, target_hidden_states):
```
Comment (weiguihua2) [inline vllm_ascend/spec_decode/llm_base_proposer.py:2004]: move to pcp_utils

**Review by weiguihua2 — APPROVED** [review]

**Review by zzzzwwjj — APPROVED** [review]

**Review by lilinsiman — APPROVED** [review]

**Review by linfeng-yuan — APPROVED** [review]
